src/Query/QueryBuilder.php

```diff
@@ -1259,6 +1267,34 @@ public function resetOrderBy(): self
         return $this;
     }
 
+    /**
+     * Adds a CTE to a select query.
+     *
+     * @param string $name         The name of the CTE.
+     * @param self   $queryBuilder The query builder to use within the CTE.
+     *
+     * @return $this This QueryBuilder instance.
+     */
+    public function with(string $name, self $queryBuilder): self
+    {
+        if (! $this->connection->getDatabasePlatform()->supportsCTEs()) {
+            throw new QueryException('CTEs are not supported by this database platform.');
+        }
+
+        if ($queryBuilder->hasCTEs()) {
+            throw new QueryException('CTEs cannot be nested.');
+        }
```

Comment:
Not sure if we need this check at all. But if we do, the check should be moved to the point where we generate the SQL statement because that query builder could be mutated after the `with()` call.